server/graphql/v1/mutations/orders.js

```diff
@@ -19,51 +19,57 @@ import {
-const ordersLimitCache = LRU({
+// Create Orders Store Object to limit FromCollective orders to up to 10 orders/hour
+const ordersLimitStore = LRU({
   max: 1000,
   maxAge: 1000 * 60 * 60, // we keep it max 1 hour
 });
 
 const debugOrder = debug('order');
 
-function checkOrdersLimit(order) {
-  const fromCollectiveIdOrUserEmailKey = get(order, 'fromCollective.id') || get(order, 'user.email');
-  const collectiveIdKey = get(order, 'collective.id');
-  // First of all, check if creation of orders has reached a limit but disabling behaviour
-  // temporarily for "test" or "circleci" ENVs(both test environments) as most tests create more than 2 orders per (fromcollective,collective) and it fails
-  if (process.env.NODE_ENV !== 'test' && process.env.NODE_ENV !== 'circleci' &&
-      fromCollectiveIdOrUserEmailKey && collectiveIdKey) {
-    const fromCollectiveValue = ordersLimitCache.get(fromCollectiveIdOrUserEmailKey);
-    // Check if FromCollective reached max limit of orders per hour
-    if (fromCollectiveValue) {
-      if (fromCollectiveValue > 10) {
-        debugOrder(`Orders Cache: ${fromCollectiveIdOrUserEmailKey} has reached the hourly limit of donations`);
-        throw new Error('Error while processing your request, please try again or contact support@opencollective.com');
-      }
-      ordersLimitCache.set(fromCollectiveIdOrUserEmailKey, fromCollectiveValue + 1);
-    } else {
-      ordersLimitCache.set(fromCollectiveIdOrUserEmailKey, 1);
-    }
-    const fromCollectiveAndCollectiveValue = ordersLimitCache.get(`${fromCollectiveIdOrUserEmailKey}_${collectiveIdKey}`);
-    // Check if pair (FromCollective, Collective) reached max limit of orders per hour
-    if (fromCollectiveAndCollectiveValue) {
-      if (fromCollectiveAndCollectiveValue > 2) {
-        debugOrder(`Orders Cache: fromCollective Id ${fromCollectiveIdOrUserEmailKey} has reached the hourly
-          limit of donations for collective id ${collectiveIdKey}`);
-        throw new Error('Error while processing your request, please try again or contact support@opencollective.com');
+function checkOrdersLimit(order, reqIp) {
+  if (process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'circleci') {
+    return;
+  }
+
+  const collectiveId = get(order, 'collective.id');
+  const userId = get(order, 'fromCollective.id') || get(order, 'user.email');
+
+  const limits = {};
+
+  // Limit on Users
+  if (userId) {
+    limits.push({ key: userId, value: 10 });
+    limits.push({ key: `${userId}_{${collectiveId}`, value: 2 });
+  }
+
+  // Limit on IPs
+  if (reqIp && !has(order, 'fromCollective.id')) {
+    limits.push({ key: reqIp, value: 10 });
+  }
+
+  for (const limit of limits) {
+    const count = ordersLimitStore.get(limit.key);
+    if (count) {
+      debugOrder(`${count} orders for limit '${limit.key}'`);
+      const limitReached = count > limit.value;
+      ordersLimitStore.set(limit.key, count + 1);
```

Comment:
We will increment the count even if the order fails, which means a user constantly retrying to order will never be able to do so. Is it intentional?

Reply:
Incrementing or not is not changing much. What matters is the 1 hour TTL in the end. The user hitting the limit will only be able to something after 1 hour, when the key expires.

Reply:
I didn't properly understand how the LRU cache worked, thanks for the explanation. Sounds good to me :+1:

Reply:
I think  there is a ttl that will solve(clean up after 1h) the problem i understood you're trying to say here. isn't it?